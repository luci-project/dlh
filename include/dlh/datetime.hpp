// Dirty Little Helper (DLH) - system support library for C/C++
// Copyright 2021-2023 by Bernhard Heinloth <heinloth@cs.fau.de>
// SPDX-License-Identifier: AGPL-3.0-or-later

#pragma once

#include <dlh/types.hpp>
#include <dlh/systypes.hpp>

/*! \brief Date and Time
 *
 * \note only Dates between 1 January 2000 and 31 December 2099 are supported
 */
struct DateTime {
	uint16_t year = 2000;  ///< Year (`0` (2000) - `99` (2099))
	uint8_t month = 1;     ///< Month (`1` - `12`)
	uint8_t day = 1;       ///< Day of Month (`1` - `31`)
	uint8_t weekday = 7;   ///< Weekday (`1` (Sun) - `7` (Sat))
	uint8_t hour = 0;      ///< Hour (`0` - `23`)
	uint8_t minute = 0;    ///< Minute (`0` - `59`)
	uint8_t second = 0;    ///< Second (`0` - `59`)
	long nanosecond = 0;   ///< nanosecond (`0` - `999999999`)


	/*! \brief Default constructor
	 */
	DateTime() {}

	DateTime(uint16_t year, uint8_t month, uint8_t day, uint8_t hour = 0, uint8_t minute = 0, uint8_t second = 0, long nanosecond = 0)
	 : year(year), month(month), day(day), weekday(0), hour(hour), minute(minute), second(second), nanosecond(nanosecond) {
		fix();
	}

	/*! \brief Construct from a timestamp (Unix epoch time)
	 */
	explicit DateTime(time_t epoch, long int nanosecond = 0) {
		from_timestamp(epoch, nanosecond);
	}

	/*! \brief Construct from a timespec struct
	 */
	explicit DateTime(const timespec & ts) {
		from_timespec(ts);
	}

	/* Copy constructor */
	constexpr DateTime(const DateTime & other)  = default;
	constexpr DateTime(DateTime && other) = default;

	/*! \brief Get name of weekday */
	const char * weekday_name(bool abbrev = false) const;

	/*! \brief Get name of month */
	const char * month_name(bool abbrev = false) const;

	/*! \brief Assign from Unix epoch time
	 */
	void from_timestamp(time_t epoch, long int nanosecond = 0);

	/*! \brief Assign from Unix timespec
	 */
	void from_timespec(const timespec & ts) {
		from_timestamp(ts.tv_sec, ts.tv_nsec);
	}

	/*! \brief Convert to Unix epoch time
	 * \return Seconds since 1 January 1970
	 */
	time_t to_timestamp() const;

	/*! \brief Convert to Unix timespec (including nanosecond)
	 */
	timespec to_timespec() const {
		timespec ts{ to_timestamp(), nanosecond };
		return ts;
	}

	/*! \brief Set to current time
	 */
	void to_current_time();

	/*! \brief Get current time
	 */
	static DateTime now();

	/*! \brief Check if date/time is valid
	 */
	bool valid() const;

	/*! \brief Fix date/time (in case of manual changes)
	 */
	void fix();


	DateTime& operator=(const time_t t) {
		from_timestamp(t);
		return *this;
	}

	DateTime& operator=(const timespec & ts) {
		from_timespec(ts);
		return *this;
	}

	constexpr DateTime& operator=(const DateTime & other) {
		this->nanosecond = other.nanosecond;
		this->second = other.second;
		this->minute = other.minute;
		this->hour = other.hour;
		this->weekday = other.weekday;
		this->day = other.day;
		this->month = other.month;
		this->year = other.year;
		return *this;
	}

	/* comparison */
	bool operator==(const time_t & other) const {
		return nanosecond == 0 && to_timestamp() == other;
	}

	bool operator==(const timespec & other) const {
		auto ts = to_timespec();
		return ts.tv_sec == other.tv_sec && ts.tv_nsec == other.tv_nsec;
	}

	bool operator==(const DateTime & other) const {
		return this->nanosecond == other.nanosecond
		    && this->second == other.second
		    && this->minute == other.minute
		    && this->hour == other.hour
		    && this->weekday == other.weekday
		    && this->day == other.day
		    && this->month == other.month
		    && this->year == other.year;
	}

	template<typename T>
	bool operator!=(const T & other) const {
		return !operator==(other);
	}


	bool operator<(const timespec & other) const {
		auto ts = to_timespec();
		return ts.tv_sec < other.tv_sec || (ts.tv_sec == other.tv_sec && ts.tv_nsec < other.tv_nsec);
	}

	bool operator<(const time_t & other) const {
		timespec other_ts = { other, 0 };
		return operator<(other_ts);
	}

	bool operator<(const DateTime & other) const {
		return operator<(other.to_timespec());
	}

	template<typename T>
	bool operator>=(const T & other) const {
		return !operator<(other);
	}

	bool operator<=(const timespec & other) const {
		auto ts = to_timespec();
		return ts.tv_sec < other.tv_sec || (ts.tv_sec == other.tv_sec && ts.tv_nsec <= other.tv_nsec);
	}

	bool operator<=(const time_t & other) const {
		timespec other_ts = { other, 0 };
		return operator<(other_ts);
	}

	bool operator<=(const DateTime & other) const {
		return operator<(other.to_timespec());
	}


	template<typename T>
	bool operator>(const T & other) const {
		return !operator<=(other);
	}
};

// Print date conform to RFC 3339 / ISO 8601
template<typename T>
T& operator<<(T &out, const DateTime &dt) {
	out.format("%04u-%02u-%02uT%02u:%02u:%02u.%06uZ", dt.year, dt.month, dt.day, dt.hour, dt.minute, dt.second, dt.nanosecond / 1000);
	return out;
}
