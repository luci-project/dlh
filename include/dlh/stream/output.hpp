#pragma once

#include <dlh/syscall.hpp>
#include <dlh/stream/string.hpp>

template<size_t BUFFERSZ>
class OutputStream : public StringStream<BUFFERSZ> {
 protected:
	int fd;

 public:
	/*! \brief Default constructor  */
	explicit OutputStream(int fd) : fd(fd) {}

	/*! \brief Destructor */
	virtual ~OutputStream() {}

	/*! \brief Flush the buffer.
	 */
	virtual void flush() override {
		size_t l = 0;
		while (l < this->_pos) {
			auto r = Syscall::write(fd, this->_bufptr + l, this->_pos - l);
			if (!r.valid())
				break;
			else
				l += r.value();
		}
		this->_pos = 0;
	}

};

extern OutputStream<1024> cout, cerr;
