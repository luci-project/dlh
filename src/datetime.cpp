// Dirty Little Helper (DLH) - system support library for C/C++
// Copyright 2021-2023 by Bernhard Heinloth <heinloth@cs.fau.de>
// SPDX-License-Identifier: AGPL-3.0-or-later

#include <dlh/datetime.hpp>

#include <dlh/syscall.hpp>
#include <dlh/math.hpp>

static const long int ns_limit = 1000000000L;

// inspired by
// https://codereview.stackexchange.com/questions/38275/convert-between-date-time-and-time-stamp-without-using-standard-library-routines
static const unsigned days[4][12] = {
	{    0,   31,   59,   90,  120,  151,  181,  212,  243,  273,  304,  334},
	{  365,  396,  424,  455,  485,  516,  547,  577,  608,  638,  669,  699},
	{  730,  761,  790,  821,  851,  882,  912,  943,  974, 1004, 1035, 1065},  // Leap Year
	{ 1096, 1127, 1155, 1186, 1216, 1247, 1277, 1308, 1339, 1369, 1400, 1430},
};

static const unsigned days4y = 365 * 4 + 1;

static const char * const weekdays[12][2] =	{
	{ "Sun", "Sunday" },
	{ "Mon", "Monday" },
	{ "Tue", "Tuesday" },
	{ "Wed", "Wednesday" },
	{ "Thu", "Thursday" },
	{ "Fri", "Friday" },
	{ "Sat", "Saturday" }
};

static const char * const months[12][2] =	{
	{ "Jan", "January" },
	{ "Feb", "February" },
	{ "Mar", "March" },
	{ "Apr", "April" },
	{ "May", "May" },
	{ "Jun", "June" },
	{ "Jul", "July" },
	{ "Aug", "August" },
	{ "Sep", "September" },
	{ "Oct", "October" },
	{ "Nov", "November" },
	{ "Dec", "December" }
};

void DateTime::from_timestamp(time_t epoch, long int nanosecond) {
	if (nanosecond != 0) {
		epoch += nanosecond / ns_limit;
		nanosecond %= ns_limit;
		if (nanosecond < 0) {
			epoch -= 1;
			nanosecond += ns_limit;
		}
	}
	this->nanosecond = nanosecond;

	second = epoch % 60;
	epoch /= 60;
	minute = epoch % 60;
	epoch /= 60;
	hour   = epoch % 24;
	epoch /= 24;

	weekday = ((epoch + 4) % 7) + 1;

	unsigned int y = 1970 + epoch / days4y * 4;
	epoch %= days4y;

	for (year = 3; year > 0 && epoch < days[year][0]; year--) {}
	for (month = 11; month > 0 && epoch < days[year][month]; month--) {}

	day = epoch - days[year][month] + 1;
	month++;
	year += y;
}

const char * DateTime::weekday_name(bool abbrev) const {
	return weekdays[Math::range(weekday, 1, 7) - 1][abbrev ? 0 : 1];
}

const char * DateTime::month_name(bool abbrev) const {
	return months[Math::range(month, 1, 12) - 1][abbrev ? 0 : 1];
}

time_t DateTime::to_timestamp() const {
	unsigned y = year - 1970;
	unsigned m = month - 1;
	unsigned d = (y / 4) * days4y + days[y % 4][m] + day - 1;
	return ((d * 24 + hour) * 60 + minute) * 60 + second;
}

void DateTime::to_current_time() {
	timespec ts;
	if (Syscall::clock_gettime(CLOCK_REALTIME, &ts).success()) {
		from_timespec(ts);
	} else {
		from_timestamp(Syscall::time());
		nanosecond = 0;
	}
}

DateTime DateTime::now() {
	DateTime dt;
	dt.to_current_time();
	return dt;
}

bool DateTime::valid()  const {
	DateTime other(to_timespec());
	return year >= 2000 && year <= 2099 && operator==(other);
}

void DateTime::fix() {
	auto ts = to_timespec();
	from_timespec(ts);
}
