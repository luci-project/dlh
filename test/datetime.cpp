#include <dlh/datetime.hpp>
#include <dlh/stream/output.hpp>

int main() {
	DateTime m0;
	DateTime m1(2021,2,18,20,55,23,421337000);
	DateTime m2(1509705000);
	DateTime m3(2012,7,37,3,136,116,1042133700);

	cout << m0 << " - " << m0.to_timestamp() << " = " << m0.valid() << endl
	     << m1 << " - " << m1.to_timestamp() << " = " << m1.valid() << endl
	     << m2 << " - " << m2.to_timestamp() << " = " << m2.valid() << endl
	     << m3 << " - " << m3.to_timestamp() << " = " << m3.valid() << endl;

	cout << m1 << (m1 <= m3 ? " was before " : " was after ") << m3 << endl;
	cout << m2 << (m2 >= m0 ? " was after " : " was before ") << m0 << endl;
	cout << m3 << (m3 > m2.to_timestamp() ? " was after " : " was before ") << m2 << endl;
	cout << m0 << (m0 <= m1.to_timespec() ? " was before " : " was after ") << m1 << endl;

	cout << m2.weekday_name() << ", " << m2.month_name() << " " << static_cast<int>(m2.day) << ", " << static_cast<int>(m2.year) << endl;
}
