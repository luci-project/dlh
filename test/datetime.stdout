2000-01-01T00:00:00.000000Z - 946684800 = true
2021-02-18T20:55:23.421337Z - 1613681723 = true
2017-11-03T10:30:00.000000Z - 1509705000 = true
2012-08-06T05:17:57.042133Z - 1344230277 = true
2021-02-18T20:55:23.421337Z was after 2012-08-06T05:17:57.042133Z
2017-11-03T10:30:00.000000Z was after 2000-01-01T00:00:00.000000Z
2012-08-06T05:17:57.042133Z was before 2017-11-03T10:30:00.000000Z
2000-01-01T00:00:00.000000Z was before 2021-02-18T20:55:23.421337Z
Friday, November 3, 2017
